### PR TITLE
[somfytahoma] Open to other portals

### PR DESCRIPTION
**Pull request by lolodomo**
Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
@octa22 : please have a look, I added the ability to choose other portals, Somfy Connexoon and Cozytouch.
Please note that for Connexoon the only difference is the URL; I applied the same session logic as for TahomaLink. I don't know if it will work.

**Comment by lolodomo** [comment]
@octa22 : please have a look.
In addition to the different portals, I added an advanced option to use or not the session cookie. I thought it was required for the Cozytouch portal but then I tried to disable it (and so to use the same code as for TahomaLink) and I was surprised to discover it was apparently working. So maybe this option is useless but it is only an advanced option, disabled by default (meaning your original code is used by default).

**Comment by lolodomo** [comment]
@octa22 : kind reminder

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 1]
```diff
@@ -233,9 +236,16 @@ public synchronized void login() {
...
                         "Received invalid data (login)");
             } else if (data.isSuccess()) {
+                if (response.getHeaders().get(HttpHeader.SET_COOKIE) != null) {
+                    List<HttpCookie> cookies = HttpCookie.parse(response.getHeaders().get(HttpHeader.SET_COOKIE));
+                    if (!cookies.isEmpty()) {
+                        cookie = cookies.get(0);
+                        logger.debug("Cookie retrieved");
+                    }
+                }
                 logger.debug("SomfyTahoma version: {}", data.getVersion());
```
Comment (octa22) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:246]: I don't think the extra cookie handling is necessary. Tahoma needs the cookie as well but the jetty HttpClient handles this out of the box (unlike HttpURLConnection which I used before)

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 2]
```diff
@@ -644,10 +656,24 @@ private String sendMethodToTahomaWithCookie(String url, HttpMethod method, Strin
...
-                .timeout(TAHOMA_TIMEOUT, TimeUnit.SECONDS).agent(TAHOMA_AGENT);
+    private Request sendRequestBuilder(String subUrl, HttpMethod method) {
+        Request request = httpClient.newRequest(getApiFullUrl(subUrl)).method(method).timeout(TAHOMA_TIMEOUT,
+                TimeUnit.SECONDS);
+        if (thingConfig.isCookieRequired()) {
+            request = request.header(HttpHeader.ACCEPT, "application/json");
+            if (!"login".equals(subUrl) && cookie != null) {
+                request = request.cookie(cookie);
+            }
+        } else {
```
Comment (octa22) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:667]: see above

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 3]
```diff
@@ -799,11 +825,11 @@ private boolean isAuthenticationChallenge(Exception ex) {
...
     public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
         super.handleConfigurationUpdate(configurationParameters);
-        if (configurationParameters.containsKey("email")) {
-            thingConfig.setEmail(configurationParameters.get("email").toString());
-        }
-        if (configurationParameters.containsKey("password")) {
-            thingConfig.setPassword(configurationParameters.get("password").toString());
+        if (configurationParameters.containsKey("email") || configurationParameters.containsKey("password")
+                || configurationParameters.containsKey("portalUrl")
+                || configurationParameters.containsKey("cookieRequired")) {
```
Comment (octa22) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:830]: see above

File: bundles/org.openhab.binding.somfytahoma/src/main/resources/OH-INF/config/config.xml [hunk 4]
```diff
@@ -19,26 +19,49 @@
...
 			<label>Password</label>
-			<description>Password for TahomaLink portal</description>
+			<description>Password for the portal</description>
+		</parameter>
+
+		<parameter name="cookieRequired" type="boolean" required="false">
+			<label>Cookie Required</label>
+			<description>Use or not a session cookie to communicate with the portal</description>
+			<default>false</default>
+			<advanced>true</advanced>
```
Comment (octa22) [inline bundles/org.openhab.binding.somfytahoma/src/main/resources/OH-INF/config/config.xml:53]: see above

**Review by octa22 — CHANGES_REQUESTED** [review]
Thanks  for the nice PR. Overall it looks good to me, I have just comment to the extra cookie handling. Not necessary in my opinion and makes the code more complex for nothing.

**Comment by lolodomo** [comment]
> I have just comment to the extra cookie handling. Not necessary in my opinion and makes the code more complex for nothing.

Are you sure that your initial code will work with any portal ?
I will try again with Cozytouch.

**Comment by lolodomo** [comment]
@octa22 : I removed the advanced setting.

**Review by octa22 — APPROVED** [review]
LGTM. Thanks!

**Review by fwolter — APPROVED** [review]

**Pull request merged**